memory_profiler.py

```diff
@@ -931,10 +980,49 @@ def wrapper(*args, **kwargs):
         return wrapper
     else:
         def inner_wrapper(f):
-            return profile(f, stream=stream, precision=precision)
+            return profile(f, stream=stream, precision=precision, backend=backend)
         return inner_wrapper
 
 
+def choose_backend():
+    """
+    Function that tries to setup backend, chosen by user, and if failed,
+    setup one of the allowable backends
+    """
+    global _backend
+    old_backend = _backend
+    backends = OrderedDict([
+        ('psutil', has_psutil),
+        ('posix', os.name == 'posix'),
+        ('tracemalloc', has_tracemalloc),
+        ('no_backend', True)
+    ])
+
+    def move_to_start(d, key):
+        """
+        Emulation of OrderedDict.move_to_end(last=False) for old versions of Python
+        """
+        items = [(key, d[key])]
+        for _key, _value in d.items():
+            if _key != key:
+                items.append((_key, _value))
+        return OrderedDict(items)
+
+    backends = move_to_start(backends, _backend)
+
+    for n_backend, is_available in backends.items():
+        if is_available:
+            _backend = n_backend
+            break
+    if _backend == 'no_backend':
+        raise NotImplementedError('Tracemalloc or psutil module is required for non-unix '
+                                  'platforms')
+    if _backend != old_backend:
+        print('{} can not be used, {} used instead'.format(old_backend, _backend))
```

Comment:
At this point Im fine dropping support for Python 2.6
On Jul 13, 2016 10:34 AM, "Sergei Lebedev" <notifications@github.com> wrote:

> In memory_profiler.py
> <https://github.com/fabianp/memory_profiler/pull/117#discussion_r70584712>
> :
>
> > +        for _key, _value in d.items():
> > +            if _key != key:
> > +                items.append((_key, _value))
> > +        return OrderedDict(items)
> > +
> > +    backends = move_to_start(backends, _backend)
> > +
> > +    for n_backend, is_available in backends.items():
> > +        if is_available:
> > +            _backend = n_backend
> > +            break
> > +    if _backend == 'no_backend':
> > +        raise NotImplementedError('Tracemalloc or psutil module is required for non-unix '
> > +                                  'platforms')
> > +    if _backend != old_backend:
> > +        print('{} can not be used, {} used instead'.format(old_backend, _backend))
>
> On second thought, OrderedDict is not in Python2.6 as well ...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/fabianp/memory_profiler/pull/117/files/fc44641d566e58f3274421eb6a1c10718dd67e64#r70584712>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAQ8h5fYIHlaerNERQUIDbof5t0cMc4oks5qVKL8gaJpZM4JKbxQ>
> .
>